src/v/pandaproxy/json/requests/fetch.h

```diff
@@ -42,20 +42,26 @@ class rjson_serialize_impl<model::record> {
       , _tpv(tpv)
       , _base_offset(base_offset) {}
 
-    void
+    bool
     operator()(::json::Writer<::json::StringBuffer>& w, model::record record) {
         w.StartObject();
         w.Key("topic");
         ::json::rjson_serialize(w, _tpv.topic);
         w.Key("key");
-        rjson_serialize_fmt(_fmt)(w, record.release_key());
+        if (!rjson_serialize_fmt(_fmt)(w, record.release_key())) {
+            throw serialize_error("Unable to serialize key value of record");
+        }
         w.Key("value");
-        rjson_serialize_fmt(_fmt)(w, record.release_value());
+        if (!rjson_serialize_fmt(_fmt)(w, record.release_value())) {
+            throw serialize_error("Unable to deserialize key value of record");
+        }
         w.Key("partition");
         ::json::rjson_serialize(w, _tpv.partition);
```

Comment:
The assertion occurred when attempting to serialize the partition.  This is because the `json::Writer` was left in a bad state when the value failed to serialize.  However, without any error checking, we didn't know it failed and kept going.